storage/dataflux/next_page.go

```diff
@@ -0,0 +1,170 @@
+		if err == iterator.Done {
+			return &nextPageResult{
+				items:          items,
+				doneListing:    true,
+				nextStartRange: "",
+				generation:     int64(0),
+			}, nil
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("iterating through objects: %w", err)
+		}
+
+		// Skip object versions already processed in the previous page to prevent duplicates.
+		if opts.query.Versions && opts.query.StartOffset == attrs.Name && attrs.Generation < opts.generation {
+			continue
+		}
+
+		if !(opts.skipDirectoryObjects && strings.HasSuffix(attrs.Name, "/")) {
+			items = append(items, attrs)
+			// Track index of the current item added to the items list.
+			itemIndex++
+		}
+
+		// If name/prefix is greater than objectLexLast, update objectLexLast and indexLexLast.
+		if objectLexLast <= attrs.Name || objectLexLast <= attrs.Prefix {
+			objectLexLast = attrs.Prefix
+			if objectLexLast <= attrs.Name {
+				objectLexLast = attrs.Name
+			}
+			// If object is added to the items list, then update indexLexLast to current item index, else set indexLexLast to -1.
+			// Setting indexLexLast to -1, indicates that the lexicographically last item is not added to items list.
+			if !(opts.skipDirectoryObjects && strings.HasSuffix(attrs.Name, "/")) {
+				indexLexLast = itemIndex
+			} else {
+				indexLexLast = -1
+			}
+		}
+
+		// If the "startoffset" value matches the name of the last object,
+		// list another page to ensure the next NextStartRange is distinct from the current one.
+		if opts.query.Versions && i == defaultPageSize-1 && attrs.Generation != int64(0) && opts.query.StartOffset == attrs.Name {
+			i = -1
+		}
+
+		// When generation value is not set, list next page if the last item is a version of previous item to prevent duplicate listing.
+		if opts.query.Versions && i == defaultPageSize-1 && attrs.Generation == int64(0) && indexLexLast > 0 && items[indexLexLast-1].Name == items[indexLexLast].Name {
+			i = -1
+		}
+	}
+
+	// Make last item as next start range.
+	nextStartRange := strings.TrimPrefix(objectLexLast, opts.query.Prefix)
+	// When the lexicographically last item is not added to items list due to skipDirectoryObjects,
+	// then set doneListing return objectLexLast as next start range.
+	if len(items) < 1 || indexLexLast == -1 {
+		return &nextPageResult{
+			items:          items,
+			doneListing:    false,
+			nextStartRange: nextStartRange,
+		}, nil
+	}
+
+	generation := int64(0)
+
+	// Remove lexicographically last item from the item list to avoid duplicate listing.
+	// Store generation of the item to be removed from the list.
+	if indexLexLast >= itemIndex {
+		generation = items[itemIndex].Generation
+		items = items[:len(items)-1]
+	} else if indexLexLast >= 0 {
+		generation = items[indexLexLast].Generation
+		items = append(items[:indexLexLast], items[indexLexLast+1:]...)
+	}
+
+	// Check if is versions is false, generation is not required.
+	if !opts.query.Versions {
+		generation = int64(0)
```

Comment:
Why is generation not required in that case?

Reply:
Since version is disabled in query, only the latest version of the object will be fetch by gcs. So generation  is not required is this case.